egress-gateway: Add IPv6 support

syedazeez337 opened the pull request:
# egress-gateway: Add IPv6 support

## Description
This PR adds IPv6 support to the egress gateway feature by introducing a new `--enable-egress-gateway` flag that enables egress gateway functionality for both IPv4 and IPv6. The existing `--enable-ipv4-egress-gateway` flag is marked as deprecated but still functional for backward compatibility.

## Changes
- Add a new `--enable-egress-gateway` flag that enables egress gateway for both IPv4 and IPv6
- Properly mark the old flag as deprecated using `flags.MarkDeprecated`
- Update the validation format for EgressIP from ipv4 to ip to support both IPv4 and IPv6 addresses
- Update documentation to remove unnecessary IPv6 examples and comments
- Update command reference documentation
- Simplify example file

## Testing
- Ran unit tests for the affected packages
- Verified that both flags can be set in the DaemonConfig struct
- Ran `make -C Documentation update-cmdref` to update the command reference

## Related Issues
Fixes #38957
Fixes #38962

```release-note
Add a new `--enable-egress-gateway` flag that enables egress gateway for both IPv4 and IPv6, deprecating the IPv4-only `--enable-ipv4-egress-gateway` flag.

zacharysarah reviewed (approved):
LGTM for docs

joestringer commented on Documentation/community/release-notes/v1.16.rst:
In this project we track release notes in two ways:
1. For overall release notes we use metadata from the PR. It's enough to have a PR like this open with a user-friendly PR name and a label `release-note/*` (I can help with that.)
2. For changes that have upgrade impact, like new flags or deprecation of existing flags, we make those changes in `Documentation/operations/upgrade.rst`. You should see existing structure there to extend for the relevant documentation.

Please remove this file.

Note that the next release is Cilium v1.18.0, estimated for release around July or so. This feature will not become part of the upstream v1.17 or earlier releases.

joestringer reviewed (changes requested):
One minor change from a @cilium/docs-structure reviewer perspective.

Please also merge all the commits together and rebase against the tip of the main branch. The merge commit seems to be causing some confusion in the pull request.

joestringer reviewed (approved):
The changes to docs look good to me :+1:

syedazeez337 committed:
egressgateway: Add IPv6 support

This commit adds IPv6 support to the egress gateway feature. It includes:

1. A new flag `--enable-egress-gateway` that enables both IPv4 and IPv6 egress gateway
   functionality, while keeping the old `--enable-ipv4-egress-gateway` flag for
   backward compatibility.
2. Updates to the CRD schema to accept both IPv4 and IPv6 addresses.
3. Validation to reject policies with mismatched IP families.
4. Support for IPv6 egress IPs in policy configuration.
5. Tests for IPv6 egress gateway functionality.

Signed-off-by: Syed Azeez <syedazeez337@gmail.com>

rgo3 commented on pkg/egressgateway/manager.go:
```diff
@@ -185,8 +185,8 @@ func NewEgressGatewayManager(p Params) (out struct {
-	// TODO: deprecate --enable-ipv4-egress-gateway, create new --enable-egress-gateway
-	if !dcfg.EnableIPv4EgressGateway {
+	// Check if either IPv4 or IPv6 egress gateway is enabled
+	if !dcfg.EnableIPv4EgressGateway && !dcfg.EnableEgressGateway {
```
I think here we can still just have one daemon config value. However you should add some code in `pkg/option/config.go`, to make sure the config value is set correctly depending on how users set the option configs flag. I would fdo this with the following steps:
1. Rename `DaemonConfig.EnableIPv4EgressGateway` to `DaemonConfig.EnableEgressGateway`
2. Set `DaemonConfig.EnableEgressGateway` to true if either of the option flags `enable-ipv4-egress-gateway` or `enable-egress-gateway` is set to true. You can access them through viper.

This way we can keep most of the code the same, we added a new option flag, deprecated the old one but nothing for existing users will change if they keep running with `enable-ipv4-egress-gateway` after an upgrade.

rgo3 commented on pkg/egressgateway/manager.go:
```diff
@@ -198,10 +198,19 @@ func NewEgressGatewayManager(p Params) (out struct {
 		return out, errors.New("egress gateway is not supported in combination with the CiliumEndpointSlice feature")
 	}
 
-	// TODO: refactor config checks for both ipv4 and ipv6, and derive whether the environment supports egress gateway policies for either protocol
```
I think for now we should keep this code as is, and keep the TODO comment. We can tackle this independently after this PR. That's why I created two separate issues. Making sure we do this correctly is important so that we don't break existing deployments. For example you suggested change could potentially break an existing user-deployment that uses ipv4 egress gateway policies, has ipv6 enabled as well but sets `dcfg.EnableIPv6Masquerade` to false deliberately. For such users cilium would suddenly fail after an upgrade, without them having changed a thing. This should be avoided.

So let's stick to deprecating the old option flag and enabling users to use the egressIP field in this PR.

rgo3 commented on pkg/egressgateway/policy.go:
```diff
@@ -359,6 +366,28 @@ func ParseCEGP(cegp *v2.CiliumEgressGatewayPolicy) (*PolicyConfig, error) {
 		}
 	}
 
+	// Check for mismatched IP families
```
I think this code could be a bit less verbose and can be run earlier in this function. I would propose to add something like the following, right after we have collected all the destination CIDRs:

```suggestion
for _, cidr := range dstCidrList {
	if cidr.Addr().Is6() && policyGwc.egressIP.IsValid() && policyGwc.egressIP.Is4() {
		return nil, fmt.Errorf("mismatched IP families: IPv4 egress IP with IPv6 destination CIDRs")
	}
	if cidr.Addr().Is4() && policyGwc.egressIP.IsValid() && policyGwc.egressIP.Is6() {
		return nil, fmt.Errorf("mismatched IP families: IPv6 egress IP with IPv4 destination CIDRs")
	}
}
```

rgo3 commented on pkg/option/config.go:
```diff
@@ -1671,6 +1674,7 @@ type DaemonConfig struct {
 
 	EnableBPFClockProbe     bool
 	EnableIPv4EgressGateway bool
```
As I suggested above, I think we can remove this daemon config parameter.

rgo3 commented on pkg/egressgateway/policy_ipv6_test.go:
Instead of adding a new ipv6 only test file, please see the `TestEgressGatewayCEGPParser` function in `manager_privileged_test.go` and add the needed test cases there.

rgo3 commented on pkg/k8s/apis/cilium.io/client/crds/v2/ciliumegressgatewaypolicies.yaml:
This file is code-generated I believe, to make any changes here you'll have to change `pkg/k8s/apis/cilium.io/v2/cegp_types.go`.

rgo3 commented on pkg/k8s/apis/cilium.io/client/crds/v2/ciliumegressgatewaypolicies.yaml:
```diff
@@ -69,9 +69,9 @@ spec:
+                      policy will use the first IP address (IPv4 or IPv6 depending on the destination CIDR) assigned to the interface with the
                       default route.
-                    format: ipv4
+                    format: ip
```
Are you certain that `ip` is a valid format parameter? I can't find it in the OpenAPI v3 spec?

rgo3 reviewed (changes requested):
Hey @syedazeez337, 

Sorry for the long delay, I'm not requested as a reviewer on this PR so it slipped of my plate. Apologies for that.

Please see my comments below for a first round of review.
